ResumableUpload.CS.Sample/ResumeableUploadSample.cs

```diff
@@ -0,0 +1,241 @@
+                        objVideosInsertRequest.ChunkSize = intHalfMBChunkSize;  // .5MB
+                        objVideosInsertRequest.ResponseReceived += videosInsertRequest_ResponseReceived; // Add event handler
+                        objVideosInsertRequest.ProgressChanged += videosInsertRequest_ProgressChanged;   // Add event handler
+                        objCancellationToken = new System.Threading.CancellationToken();
+                        Globals.ResumeRetries = 0;
+                        try
+                        {
+                            await objVideosInsertRequest.UploadAsync(objCancellationToken);
+                        }
+                        catch
+                        {
+                            // Error handling is done in videosInsertRequest_ProgressChanged() Event
+                        }
+                        bool bResumeAsyncWasLaunched = false;
+                        do
+                        {
+                            Thread.Sleep(3000); // Allow time for ProgressChanged() Event to execute and to allow time for a server or network issue to resolve
+                            bResumeAsyncWasLaunched = false;
+                            if (IsResumeable())
+                            {
+                                // Try a ResumeAsync
+                                Globals.ResumeRetries += 1;
+                                bResumeAsyncWasLaunched = true;
+                                try
+                                {
+                                   await objVideosInsertRequest.ResumeAsync(objCancellationToken);
+                                }
+                                catch
+                                {
+                                    // Error handling is done in videosInsertRequest_ProgressChanged() Event
+                                }
+                            }
+                        } while (bResumeAsyncWasLaunched && IsResumeable());
+                        objFileStream.Close();
+                    }
+                }
+            }
+        }
+        /// <summary>Determines if a .ResumeAsync should be attempted.</summary>
+        /// <returns>True then .ResumeAsync should be attempted.</returns>
+        /// <remarks>When Globals.ResumeRetries==int32.MaxValue, the upload has completed successfully and IsResumeable() will return false.</remarks>
+        private static bool IsResumeable()
+        {
+            return ((Globals.ResumeRetries < MAX_RESUME_RETRIES) && (!objCancellationToken.IsCancellationRequested));
+        }
+        /// <summary>
+        /// The event handler called by the .NET Client Library to indicate the current status of the upload.
+        /// </summary>
+        /// <param name="objUploadStatus"></param>
+        static void videosInsertRequest_ProgressChanged(Google.Apis.Upload.IUploadProgress objUploadStatus)
```

Comment:
It is probably worth-while to have "using Google.Apis.Upload;" at the top of the file so that you can just have the type of the parameter be `IUploadProgress`.